File: tier1_foundation/fast_slow_pointers/problems/p202_happy_number.mjs
/**
 * P202: Happy Number (Easy)
 * https://leetcode.com/problems/happy-number/
 * Topics: Hash Table, Math, Two Pointers
 * 
 * Write an algorithm to determine if a number n is happy.
 * A happy number is a number defined by the following process:
 * Return true if n is a happy number, and false if not.
 * 
 * Example 1:
 *     Input: n = 19
 *     Output: true
 *     Explanation:
 *     12 + 92 = 82
 *     82 + 22 = 68
 *     62 + 82 = 100
 *     12 + 02 + 02 = 1
 * 
 * Example 2:
 *     Input: n = 2
 *     Output: false
 * 
 * Constraints:
 *     - 1 <= n <= 231 - 1
 * 
 * Template (python3):
 *     class Solution:
 *         def isHappy(self, n: int) -> bool:
 * 
 * Hint: Use fast and slow pointers on the sequence of sum-of-squared-digits.
 */
function getNext(num)
  let total = 0;
  while (num > 0)
    const digit = num % 10;
    total += digit * digit;
    num = Math.floor(num / 10);
  }
  return total; {
    throw new Error("NotImplementedError");
}

function isHappy(n)
  let slow = n;
  let fast = getNext(n);
  while (fast !== 1 && slow !== fast)
    slow = getNext(slow);
    fast = getNext(getNext(fast));
  }
  return fast === 1; {
    throw new Error("NotImplementedError");
}

function solve(input)
  return isHappy(input); {
    throw new Error("NotImplementedError");
}

// --- tests ---
const tests = [
  { label: "example 1", input: 19, expected: true },
  { label: "example 2", input: 2, expected: false },
  { label: "already happy", input: 1, expected: true },
  { label: "small happy number", input: 7, expected: true },
  { label: "enters 4-16 cycle", input: 4, expected: false },
  { label: "power of 10", input: 100, expected: true },
  { label: "large unhappy number", input: 999999999, expected: false },
];
let passed = 0;
for (let i = 0; i < tests.length; i++) {
  const t = tests[i];
  const got = solve(t.input);
  if (JSON.stringify(got) === JSON.stringify(t.expected)) {
    passed++;
    console.log(`  Test ${i + 1} (${t.label}): PASS`);
  } else {
    console.log(`  Test ${i + 1} (${t.label}): FAIL`);
    console.log(`    Expected: ${JSON.stringify(t.expected)}\n    Got:      ${JSON.stringify(got)}`);
  }
}
console.log(`\n  ${passed}/${tests.length} passed`);
process.exit(passed === tests.length ? 0 : 1);

File: tier1_foundation/fast_slow_pointers/solutions/p202_happy_number.rs
/*
 * P202: Happy Number (Easy)
 * https://leetcode.com/problems/happy-number/
 * Topics: Hash Table, Math, Two Pointers
 *
 * Write an algorithm to determine if a number n is happy.
 * A happy number is a number defined by the following process:
 *
 * 1. Starting with any positive integer, replace the number by the sum of the squares of its digits.
 * 2. Repeat the process until the number equals 1 (where it will stay), or it loops endlessly in a cycle which does not include 1.
 * 3. Those numbers for which this process ends in 1 are happy.
 * Return true if n is a happy number, and false if not.
 *
 * Example 1:
 *     Input: n = 19
 *     Output: true
 *     Explanation:
 *     1^2 + 9^2 = 82
 *     8^2 + 2^2 = 68
 *     6^2 + 8^2 = 100
 *     1^2 + 0^2 + 0^2 = 1
 *
 * Example 2:
 *     Input: n = 2
 *     Output: false
 *
 * Constraints:
 *     - 1 <= n <= 2^31 - 1
 *
 * Template (python3):
 *     class Solution:
 *         def isHappy(self, n: int) -> bool:
 *
 * Hint: Use fast and slow pointers on the sequence of sum-of-squared-digits.
 */

use wasm_libs::*;

fn get_next(mut num: i32) -> i32 {
    let mut total = 0;
    while num > 0 {
        let digit = num % 10;
        total += digit * digit;
        num /= 10;
    }
    total
}

fn solve(n: i32) -> bool {
    let mut slow = n;
    let mut fast = get_next(n);
    while fast != 1 && slow != fast {
        slow = get_next(slow);
        fast = get_next(get_next(fast));
    }
    fast == 1
}

fn main() {
    let n = read_int();
    write_bool(solve(n));
    std::process::exit(0);
}
